File: frontend/lib/models/character.dart
// --- 데이터 모델 클래스 ---

// 스탯 정보 모델
class Stat {
  int strength;      // 근력 (빨간색)
  int intelligence;  // 지능 (파란색)
  int agility;       // 민첩 (구 Stamina) (초록색)
  int defense;       // 방어력
  int luck;          // 운
  int happiness;     // 행복도 (핑크색)
  int health;        // 현재 체력
  int exp;           // 현재 경험치
  int level;         // 레벨
  int unused_points; // 미사용 스탯 포인트 (분배 가능)

  Stat({
    required this.strength,
    required this.intelligence,
    required this.agility,
    required this.defense,
    required this.luck,
    required this.happiness,
    required this.health,
    this.exp = 0,
    this.level = 1,
    this.unused_points = 0,
  });

  // JSON 파싱 (서버 -> 앱)
  factory Stat.fromJson(Map<String, dynamic> json) {
    return Stat(
      strength: json['strength'] ?? 0,
      intelligence: json['intelligence'] ?? 0,
      agility: json['agility'] ?? 0,
      defense: json['defense'] ?? 10, // Def default
      luck: json['luck'] ?? 5,        // Luck default
      happiness: json['happiness'] ?? 0,
      health: json['health'] ?? 100,
      exp: json['exp'] ?? 0,
      level: json['level'] ?? 1,
      unused_points: json['unused_points'] ?? 0,
    );
  }
}

// 캐릭터 정보 모델
class Character {
  final int id;
  final int userId;
  final String name;
  String imageUrl; // 이미지 경로 (상태에 따라 변경됨)
  final String petType; // 반려동물 종류 (dog, cat 등) - [New]
  final List<int> learnedSkills; // 습득한 기술 ID 목록 - [New]
  final Stat? stat; // 연관된 스탯 객체

  Character({
    required this.id,
    required this.userId,
    required this.name,
    this.imageUrl = 'assets/images/characters/닌자옷.png', 
    this.petType = 'dog', // 기본값
    this.learnedSkills = const [],
    this.stat,
  });

  // JSON 파싱 (서버 -> 앱)
  factory Character.fromJson(Map<String, dynamic> json) {
    return Character(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      name: json['name'] ?? 'Unknown',
      // 이미지 URL이 없으면 기본 이미지 사용
      imageUrl: json['image_url'] ?? 'assets/images/characters/닌자옷.png',
      petType: json['pet_type'] ?? 'dog', // 서버 데이터 반영
      learnedSkills: json['learned_skills'] != null 
          ? List<int>.from(json['learned_skills']) 
          : [],
      // 중첩된 JSON 처리
      stat: (json['stats'] ?? json['stat']) != null ? Stat.fromJson(json['stats'] ?? json['stat']) : null,
    );
  }
}

File: backend/app/db/models/character.py
from sqlalchemy import Integer, String, ForeignKey, DateTime, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import JSONB
from datetime import datetime
from typing import Optional
from app.db.database import Base

# --- 캐릭터(Character) 모델 ---
class Character(Base):
    __tablename__ = "characters"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), unique=True) # 유저당 하나의 캐릭터 (1:1)
    name: Mapped[str] = mapped_column(String, index=True) # 캐릭터 이름
    status: Mapped[str] = mapped_column(String, default="normal") # 상태 (예: normal, hungry, sleepy)
    pet_type: Mapped[str] = mapped_column(String, default="dog") # 반려동물 종류 (dog, cat 등)
    learned_skills: Mapped[list[int]] = mapped_column(JSONB, default=[]) # 습득한 기술 ID 리스트
    
    # 관계 설정 (Relationships)
    user: Mapped["User"] = relationship("User", back_populates="character")
    
    # 1:1 관계 - uselist=False
    stat: Mapped["Stat"] = relationship("Stat", back_populates="character", uselist=False, cascade="all, delete-orphan")
    
    # 1:N 관계 - 행동 로그
    action_logs: Mapped[list["ActionLog"]] = relationship("ActionLog", back_populates="character", cascade="all, delete-orphan")


# --- 스탯(Stat) 모델 ---
class Stat(Base):
    __tablename__ = "stats"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    character_id: Mapped[int] = mapped_column(ForeignKey("characters.id"), unique=True)
    
    # 주요 능력치
    health: Mapped[int] = mapped_column(Integer, default=100)      # 체력
    happiness: Mapped[int] = mapped_column(Integer, default=50)    # 행복도
    exp: Mapped[int] = mapped_column(Integer, default=0)           # 경험치
    level: Mapped[int] = mapped_column(Integer, default=1)         # 레벨
    strength: Mapped[int] = mapped_column(Integer, default=10)     # 근력
    intelligence: Mapped[int] = mapped_column(Integer, default=10) # 지능
    stamina: Mapped[int] = mapped_column(Integer, default=10)      # 지구력
    defense: Mapped[int] = mapped_column(Integer, default=10)      # 방어력
    luck: Mapped[int] = mapped_column(Integer, default=5)          # 운
    personality: Mapped[str] = mapped_column(String, default="기본") # 성격
    condition: Mapped[int] = mapped_column(Integer, default=100)   # 컨디션 (0-100)
    
    # 훈련 성공 시 획득하는 분배 가능 포인트
    unused_points: Mapped[int] = mapped_column(Integer, default=5)

    character: Mapped["Character"] = relationship("Character", back_populates="stat")


# --- 행동 로그(ActionLog) 모델 ---
class ActionLog(Base):
    __tablename__ = "action_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    character_id: Mapped[int] = mapped_column(ForeignKey("characters.id"))
    
    action_type: Mapped[str] = mapped_column(String) # 수행한 행동 유형 (예: "playing_fetch")
    yolo_result_json: Mapped[dict] = mapped_column(JSONB, nullable=True) # AI 분석 결과 원본 (PostgreSQL JSONB 저장)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow) # 생성 시간

    character: Mapped["Character"] = relationship("Character", back_populates="action_logs")